### PR TITLE
[Drop-In UI] Separate binders for action buttons and add for other extendable buttons

### PR DESCRIPTION
**Pull request by tomaszrybakiewicz**
Closes NAVAND-711

### TL;DR;
This PR splits `ActionButtonBinder` into smaller binders that allow the injection of custom binders for Compass, Camera Mode, Audio Guidance and Recenter buttons. It also introduces two binders for Info Panel Route Preview and Start Navigation buttons.

| Current Binder Setup | New Binders Setup  |
|:---:|:---:|
|drop in binder diagram|drop in binder diagram2|

### Description

- Introduced UIBinder override for map action buttons.
- Refactored ActionButtonBinder:
  - renamed to ActionButtonsBinder
  - converted it to an abstract class (similar to InfoPanelBinder)
  - updated bind method logic
  - provided default ActionButtonsBinder implementation: MapboxActionButtonsBinder 
  - Restructured default action buttons XML layout to allow independent button binding
- Added UIBinders for Route Preview and Start Navigation Buttons.

### TODO
- [x] CHANGELOG entry
- [x] Add examples to qa-test-app

### CODE REVIEW

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewBinderCustomization.kt [hunk 1]
```diff
@@ -100,10 +113,38 @@ class ViewBinderCustomization {
...
     /**
      * Customize the Action Buttons by providing your own [UIBinder].
-     * Use [UIBinder.USE_DEFAULT] to reset to default.
+     * Use [UIBinder.USE_DEFAULT] to reset to use default [ActionButtonsBinder] implementation.
      */
     var actionButtonsBinder: UIBinder? = null
 
+    /**
+     * Customize the Compass Action Button by providing your own [UIBinder]
+     * when using [actionButtonsBinder] with [ActionButtonsBinder] implementation.
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewBinderCustomization.kt:122]:
```suggestion
     * when using [actionButtonsBinder] with [ActionButtonsBinder.defaultBinder] implementation.
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewBinderCustomization.kt [hunk 2]
```diff
@@ -100,10 +113,38 @@ class ViewBinderCustomization {
...
+     * Use [UIBinder.USE_DEFAULT] to reset to default.
+     */
+    var actionCompassButtonBinder: UIBinder? = null
+
+    /**
+     * Customize the Camera Mode Action Button by providing your own [UIBinder]
+     * when using [actionButtonsBinder] with [ActionButtonsBinder] implementation.
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewBinderCustomization.kt:129]:
```suggestion
     * when using [actionButtonsBinder] with [ActionButtonsBinder.defaultBinder] implementation.
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewBinderCustomization.kt [hunk 3]
```diff
@@ -100,10 +113,38 @@ class ViewBinderCustomization {
...
+     * Use [UIBinder.USE_DEFAULT] to reset to default.
+     */
+    var actionCameraModeButtonBinder: UIBinder? = null
+
+    /**
+     * Customize the Toggle Audio Guidance Action Button by providing your own [UIBinder]
+     * when using [actionButtonsBinder] with [ActionButtonsBinder] implementation.
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewBinderCustomization.kt:136]:
```suggestion
     * when using [actionButtonsBinder] with [ActionButtonsBinder.defaultBinder] implementation.
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewBinderCustomization.kt [hunk 4]
```diff
@@ -100,10 +113,38 @@ class ViewBinderCustomization {
...
+     * Use [UIBinder.USE_DEFAULT] to reset to default.
+     */
+    var actionToggleAudioButtonBinder: UIBinder? = null
+
+    /**
+     * Customize the Recenter Camera Action Button by providing your own [UIBinder]
+     * when using [actionButtonsBinder] with [ActionButtonsBinder] implementation.
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/ViewBinderCustomization.kt:143]:
```suggestion
     * when using [actionButtonsBinder] with [ActionButtonsBinder.defaultBinder] implementation.
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/ActionButtonsBinder.kt [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    internal var context: NavigationViewContext? = null
+    internal var customButtons: List<ActionButtonDescription> = emptyList()
+
+    /**
+     * Create layout that will host action buttons containers.
+     *
+     * @param layoutInflater The LayoutInflater service.
+     * @param root Parent view that provides a set of LayoutParams values for root of the returned hierarchy.
+     *
+     * @return ViewGroup that can host both header and content layouts.
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/ActionButtonsBinder.kt:38]:
```suggestion
     * @return ViewGroup that can host action button layout.
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/ActionButtonsBinder.kt [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+         */
+        fun defaultBinder(): ActionButtonsBinder = MapboxActionButtonsBinder()
+    }
+}
+
+/**
+ * Default Action Buttons Binder implementation.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+internal class MapboxActionButtonsBinder : ActionButtonsBinder() {
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/ActionButtonsBinder.kt:191]: Let's move this out to a separate class

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/AudioGuidanceButtonBinder.kt [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    override fun bind(viewGroup: ViewGroup): MapboxNavigationObserver {
+        return reloadOnChange(context.styles.audioGuidanceButtonParams) { params ->
+            val button = MapboxAudioGuidanceButton(viewGroup.context, null, 0, params.style)
+            viewGroup.removeAllViews()
+            viewGroup.addView(button, params.layoutParams)
+
+            AudioGuidanceButtonComponent(button) {
+                AudioComponentContractImpl(context.viewModel.viewModelScope, context.store)
+            }
+        }
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/AudioGuidanceButtonBinder.kt:28]: Why not do the same as you are doing here https://github.com/mapbox/mapbox-navigation-android/pull/6490/files#diff-bfe82a9aa5902eb8b53e4b16729122d2707e7ee830fc65c37ea592410fcc09f1R16-R21

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CameraModeButtonBinder.kt [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            viewGroup.removeAllViews()
+            viewGroup.addView(button, params.layoutParams)
+
+            CameraModeButtonComponent(button) {
+                CameraModeButtonComponentContractImpl(
+                    context.viewModel.viewModelScope,
+                    context.store
+                )
+            }
+        }
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CameraModeButtonBinder.kt:31]: Why not do the same as you are doing here https://github.com/mapbox/mapbox-navigation-android/pull/6490/files#diff-bfe82a9aa5902eb8b53e4b16729122d2707e7ee830fc65c37ea592410fcc09f1R16-R21

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CompassButtonBinder.kt [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+            context.styles.compassButtonParams,
+            context.mapViewOwner.mapViews
+        ) { params, mapView ->
+            val button = MapboxExtendableButton(viewGroup.context, null, 0, params.style)
+            viewGroup.removeAllViews()
+            viewGroup.addView(button, params.layoutParams)
+
+            CompassButtonComponent(button, mapView, params.enabled)
+        }
+    }
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CompassButtonBinder.kt:28]: Why not do the same as you are doing here https://github.com/mapbox/mapbox-navigation-android/pull/6490/files#diff-bfe82a9aa5902eb8b53e4b16729122d2707e7ee830fc65c37ea592410fcc09f1R16-R21

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/RecenterButtonBinder.kt [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    override fun bind(viewGroup: ViewGroup): MapboxNavigationObserver {
+        return reloadOnChange(context.styles.recenterButtonParams) { params ->
+            val button = MapboxExtendableButton(viewGroup.context, null, 0, params.style)
+            viewGroup.removeAllViews()
+            viewGroup.addView(button, params.layoutParams)
+
+            RecenterButtonComponent(button) {
+                RecenterButtonComponentContractImpl(context.viewModel.viewModelScope, context.store)
+            }
+        }
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/RecenterButtonBinder.kt:28]: Why not do the same as you are doing here https://github.com/mapbox/mapbox-navigation-android/pull/6490/files#diff-bfe82a9aa5902eb8b53e4b16729122d2707e7ee830fc65c37ea592410fcc09f1R16-R21

File: qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt [hunk 11]
```diff
@@ -88,71 +84,21 @@ import com.mapbox.navigation.qa_test_app.dump.NavigationViewApiDumpInterceptor
...
 import com.mapbox.navigation.ui.maps.NavigationStyles
-import com.mapbox.navigation.ui.maps.route.RouteLayerConstants
-import com.mapbox.navigation.ui.maps.route.arrow.model.RouteArrowOptions
-import com.mapbox.navigation.ui.maps.route.line.model.MapboxRouteLineOptions
-import com.mapbox.navigation.ui.maps.route.line.model.RouteLineColorResources
-import com.mapbox.navigation.ui.maps.route.line.model.RouteLineResources
 import com.mapbox.navigation.utils.internal.toPoint
 
 @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
 class MapboxNavigationViewCustomizedActivity : DrawerActivity() {
```
Comment (tomaszrybakiewicz) [inline qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/customnavview/MapboxNavigationViewCustomizedActivity.kt:91]: NOTE: I've restructured this activity to make it easier for us to add new customization options.
I've prefixed all switch/spinner handler methods with `toggleXXXX and grouped them with code folding regions.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/ActionButtonsBinder.kt (hunk 6 above)
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/ActionButtonsBinder.kt:191]: I assume you mean file.

Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/ActionButtonsBinder.kt:191]: Yep

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/AudioGuidanceButtonBinder.kt (hunk 7 above)
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/AudioGuidanceButtonBinder.kt:28]: I see why it can't be done because the `AudioGuidanceButtonComponent` lives outside `Drop-In UI`. Same applies to all the others below.

**Comment by abhishek1508** [comment]
A couple of other things that comes to mind:
- Are you planning to remove `MapboxExtendableButtonParams`, now that we have binders for each of these buttons? Based on the original ticket  _This PR tracks removing MapboxExtendableButtonParams in favor of separating action buttons into individual binders._ 
We can revisit this. I am thinking over it again and I feel like it's good to have this class. Consider the use case where a customer is fine with Mapbox implementation of action buttons. All they want to do is make them circular. They would have to write their own binder and layout files to make them circular. Consider another use case, where a user wants to show compass button which is hidden by default. They would have to write their own binder just to make it visible, whereas they can just toggle `MapboxExtendableButtonParams.enabled` to show/hide it while using `MapboxActionButtonBinder`. 
Wdyt?
- If we were to keep `MapboxExtendableButtonParams`, just like we do the following in `CompassButtonComponent`
```
override fun onAttached(mapboxNavigation: MapboxNavigation) {
        super.onAttached(mapboxNavigation)

        compassButton.isVisible = enabled
```
we would have to make sure we handle the visibility for `StartNavigationButtonComponent`, `EndNavigationButtonComponent`, `RoutePreviewButtonComponent`, `RecenterButtonComponent`, `AudioGuidanceButtonComponent`, `CameraModeButtonComponent`. We need not do it in this PR, but capturing it here so we don't miss it.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CameraModeButtonBinder.kt (hunk 8 above)
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CameraModeButtonBinder.kt:31]: Because UIComponents should not be responsible for creation/re-creation of the views. That's Binders responsibility.
The `RoutePreviewButtonComponent` has been wrongfully refactored some time ago to handle button re-creation (and in fact violate the above rule) when `MapboxExtendableButtonParams` was introduced. 
Since refactoring of the `RoutePreviewButtonComponent` is outside the scope of this PR, I've limited my changes to just the introduction of `InfoPanelRoutePreviewButtonBinder` that does nothing but return the component.
In a follow-up we should address this discrepancy and migrate view recreation logic from UIComponents to Binders.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CompassButtonBinder.kt (hunk 9 above)
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CompassButtonBinder.kt:28]: Answered here https://github.com/mapbox/mapbox-navigation-android/pull/6490#discussion_r999887338

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/RecenterButtonBinder.kt (hunk 10 above)
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/RecenterButtonBinder.kt:28]: Answered here https://github.com/mapbox/mapbox-navigation-android/pull/6490#discussion_r999887338

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CameraModeButtonBinder.kt (hunk 8 above)
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/actionbutton/CameraModeButtonBinder.kt:31]: 👍🏻

**Comment by tomaszrybakiewicz** [comment]
@abhishek1508 Thanks for the review. I've applied all your suggestions. This one is ready to go.
Let me know if anything else needs to change.

**Review by abhishek1508 — APPROVED** [review]
💯
